File: shared/deadline/DeadlineBudget.ts
import type { IClock } from "../clock/IClock";

/**
 * Assessment-imposed hard ceiling: no synchronous request may exceed this, including
 * S3/AI/third-party handshakes. Not configurable — it is a contract with the caller,
 * not a tuning knob.
 */
export const DEADLINE_HARD_CEILING_MS = 45_000;

/** Internal default target, below the hard ceiling, leaving headroom for cleanup and
 * response serialization. Overridable via config (DEADLINE_TARGET_MS) but validated
 * against the ceiling in DeadlineBudget.start. */
export const DEFAULT_DEADLINE_TARGET_MS = 35_000;

export interface DeadlineBudgetOptions {
  /** From `context.getRemainingTimeInMillis()` at handler entry. */
  remainingLambdaTimeMs: number;
  /** Internal target in ms; must not exceed DEADLINE_HARD_CEILING_MS. */
  targetMs?: number;
  clock: IClock;
}

/**
 * The single deadline primitive every handler/service/adapter uses. Created once at
 * handler entry from Lambda's actual remaining execution time, then threaded down
 * explicitly through every call that does I/O — never re-derived, never a fresh
 * ad-hoc timeout at the call site.
 */
export class DeadlineBudget {
  readonly startedAtMs: number;
  readonly deadlineAtMs: number;
  private readonly clock: IClock;

  private constructor(startedAtMs: number, deadlineAtMs: number, clock: IClock) {
    this.startedAtMs = startedAtMs;
    this.deadlineAtMs = deadlineAtMs;
    this.clock = clock;
  }

  static start(options: DeadlineBudgetOptions): DeadlineBudget {
    const targetMs = options.targetMs ?? DEFAULT_DEADLINE_TARGET_MS;
    if (targetMs > DEADLINE_HARD_CEILING_MS) {
      throw new RangeError(
        `Deadline target (${targetMs}ms) must not exceed the hard ceiling (${DEADLINE_HARD_CEILING_MS}ms)`,
      );
    }
    if (targetMs <= 0) {
      throw new RangeError(`Deadline target must be positive, got ${targetMs}ms`);
    }

    // The Lambda invocation itself may have less time left than our internal target
    // (e.g. a retry deeper into an already-long invocation) — always take the smaller.
    const effectiveMs = Math.max(0, Math.min(targetMs, options.remainingLambdaTimeMs));
    const startedAtMs = options.clock.now();
    return new DeadlineBudget(startedAtMs, startedAtMs + effectiveMs, options.clock);
  }

  /** Milliseconds left until the internal deadline. Never negative. */
  remainingMs(): number {
    return Math.max(0, this.deadlineAtMs - this.clock.now());
  }

  /** Milliseconds elapsed since the budget was started. */
  elapsedMs(): number {
    return this.clock.now() - this.startedAtMs;
  }

  isExpired(): boolean {
    return this.remainingMs() <= 0;
  }

  /**
   * Timeout budget for a single outbound call: remaining time minus a reserve kept
   * for cleanup/response serialization, capped at an optional per-call maximum.
   * Never negative — callers should treat 0 as "do not attempt this call."
   */
  forCall(reserveMs: number, perCallMaxMs?: number): number {
    const available = this.remainingMs() - reserveMs;
    const capped = perCallMaxMs === undefined ? available : Math.min(available, perCallMaxMs);
    return Math.max(0, capped);
  }

  /** True if there is enough budget left to attempt another call plus the reserve. */
  canAttempt(reserveMs: number): boolean {
    return this.remainingMs() > reserveMs;
  }
}

File: src/Gweb.Shared/Deadline/DeadlineBudget.cs
using Gweb.Shared.Clock;

namespace Gweb.Shared.Deadline;

/// <summary>
/// The single deadline primitive every handler/service/adapter uses. Created once at
/// handler entry from Lambda's actual remaining execution time, then threaded down
/// explicitly through every call that does I/O — never re-derived, never a fresh
/// ad-hoc timeout at the call site.
/// </summary>
public sealed class DeadlineBudget
{
    /// <summary>
    /// Assessment-imposed hard ceiling: no synchronous request may exceed this,
    /// including S3/AI/third-party handshakes. Not configurable — it is a contract
    /// with the caller, not a tuning knob.
    /// </summary>
    public const long HardCeilingMs = 45_000;

    /// <summary>
    /// Internal default target, below the hard ceiling, leaving headroom for cleanup
    /// and response serialization. Overridable via config (DEADLINE_TARGET_MS) but
    /// validated against the ceiling in <see cref="Start"/>.
    /// </summary>
    public const long DefaultTargetMs = 35_000;

    private readonly IClock _clock;

    public long StartedAtMs { get; }
    public long DeadlineAtMs { get; }

    private DeadlineBudget(long startedAtMs, long deadlineAtMs, IClock clock)
    {
        StartedAtMs = startedAtMs;
        DeadlineAtMs = deadlineAtMs;
        _clock = clock;
    }

    public static DeadlineBudget Start(long remainingLambdaTimeMs, IClock clock, long? targetMs = null)
    {
        var target = targetMs ?? DefaultTargetMs;
        if (target > HardCeilingMs)
        {
            throw new ArgumentOutOfRangeException(
                nameof(targetMs), target, $"Deadline target ({target}ms) must not exceed the hard ceiling ({HardCeilingMs}ms)");
        }
        if (target <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(targetMs), target, "Deadline target must be positive");
        }

        // The Lambda invocation itself may have less time left than our internal target
        // (e.g. a retry deeper into an already-long invocation) — always take the smaller.
        var effectiveMs = Math.Max(0, Math.Min(target, remainingLambdaTimeMs));
        var startedAtMs = clock.NowMs();
        return new DeadlineBudget(startedAtMs, startedAtMs + effectiveMs, clock);
    }

    /// <summary>Milliseconds left until the internal deadline. Never negative.</summary>
    public long RemainingMs() => Math.Max(0, DeadlineAtMs - _clock.NowMs());

    /// <summary>Milliseconds elapsed since the budget was started.</summary>
    public long ElapsedMs() => _clock.NowMs() - StartedAtMs;

    public bool IsExpired() => RemainingMs() <= 0;

    /// <summary>
    /// Timeout budget for a single outbound call: remaining time minus a reserve kept
    /// for cleanup/response serialization, capped at an optional per-call maximum.
    /// Never negative — callers should treat 0 as "do not attempt this call."
    /// </summary>
    public long ForCall(long reserveMs, long? perCallMaxMs = null)
    {
        var available = RemainingMs() - reserveMs;
        var capped = perCallMaxMs is null ? available : Math.Min(available, perCallMaxMs.Value);
        return Math.Max(0, capped);
    }

    /// <summary>True if there is enough budget left to attempt another call plus the reserve.</summary>
    public bool CanAttempt(long reserveMs) => RemainingMs() > reserveMs;
}
